starboard/raspi/shared/launcher.py

```diff
@@ -188,19 +193,38 @@ def _PexpectSpawnAndConnect(self, command):
           self._PexpectSendLine('echo ' + Launcher._SSH_LOGIN_SIGNAL)
           i = self.pexpect_process.expect([Launcher._SSH_LOGIN_SIGNAL])
           break
-      except pexpect.TIMEOUT:
+      except (pexpect.TIMEOUT, OSError, pexpect.ExceptionPexpect) as err:
         if self.shutdown_initiated.is_set():
           return
         retry_count += 1
         # Check if the max retry count has been exceeded. If it has, then
         # re-raise the timeout exception.
         if retry_count > Launcher._PEXPECT_PASSWORD_TIMEOUT_MAX_RETRIES:
-          raise
+          raise RuntimeError(f'Failed after {retry_count} retries') from err
+
+  def _PexpectSpawnAndConnect(self, command):
+    """Outer retry wrapper"""
+    retry_count = 0
+    try:
+      self._PexpectSpawnAndConnect_inner(command)
+    except RuntimeError as err:
+      retry_count += 1
+      if retry_count > Launcher._PEXPECT_OUTER_RETRIES:
+        raise RuntimeError(
+            f'Outer pexpect retry failed after {retry_count}') from err
 
   def _PexpectSendLine(self, cmd):
     """Send lines to Pexpect and record the last command for logging purposes"""
-    self.last_run_pexpect_cmd = cmd
-    self.pexpect_process.sendline(cmd)
+    retries = 5
```

Comment:
Again, I don't see a loop so I'm not sure how this would get to 0?

Reply:
Because loop is missing. Added